editor/src/core/model/element-metadata-utils.ts

```diff
@@ -1362,6 +1370,78 @@ export const MetadataUtils = {
   },
 }
 
+function fillSpyOnlyMetadataWithFramesFromChildren(
+  fromSpy: ElementInstanceMetadataMap,
+  fromDOM: ElementInstanceMetadataMap,
+) {
+  const childrenInDomCache: { [pathStr: string]: Array<ElementInstanceMetadata> } = {}
+
+  const findChildrenInDomRecursively = (pathStr: string): Array<ElementInstanceMetadata> => {
+    const existing = childrenInDomCache[pathStr]
+
+    if (existing != null) {
+      return existing
+    }
+
+    const spyElem = fromSpy[pathStr]
+    const childrenFromSpy = MetadataUtils.getChildren(fromSpy, spyElem.elementPath)
+    const childrenFromDom = MetadataUtils.getChildren(fromDOM, spyElem.elementPath)
+    const childrenNotInDom = childrenFromSpy.filter((childNotInDom) =>
+      childrenFromDom.every(
+        (childInDom) => !EP.pathsEqual(childNotInDom.elementPath, childInDom.elementPath),
+      ),
+    )
+    const recursiveChildren = childrenNotInDom.flatMap((c) => {
+      return findChildrenInDomRecursively(EP.toString(c.elementPath))
+    })
+    const children = [...childrenFromDom, ...recursiveChildren]
+
+    childrenInDomCache[pathStr] = children
+
+    return children
+  }
+
+  // those elements which are not in the dom need calculated localFrame and globalFrame
+  // from their children (or deeper descendants)
+  const elementsWithoutDomMetadata = Object.keys(fromSpy).filter((p) => fromDOM[p] == null)
+
+  const workingElements: ElementInstanceMetadataMap = {}
+
+  fastForEach(elementsWithoutDomMetadata, (pathStr) => {
+    const spyElem = fromSpy[pathStr]
+    const children = findChildrenInDomRecursively(pathStr)
+    if (children.length === 0) {
+      return
+    }
+    let workingMetadata: ElementInstanceMetadata = { ...spyElem }
+
+    fastForEach(children, (childMetadata) => {
+      const globalFrame =
+        workingMetadata.globalFrame != null
+          ? boundingRectangle(
```

Comment:
I believe these calculations will be wrong if we're falling back to using `zeroCanvasRect` at any point, since that will mean the bounding rect will always include the point `(0, 0)`, which probably won't be the case most of the time

Reply:
Do any of the tests cover this case?

Reply:
The tests cover this case, and it is not a problem, because globalFrame is always null, but childMetadata is never null, because we explicitly choose the children in a way that they have data from the dom walker.
However, you are right that we should fallback to null instead of zero rect.

Reply:
Better and way simpler solution: f569e233e38e07734beded4ab85186b194c3fe8b
I realized there is a function to calculate the bounding box for an array of rectangles (or nulls)

Reply:
Perfect 👍

Reply:
Oh the implementation in https://github.com/concrete-utopia/utopia/commit/f569e233e38e07734beded4ab85186b194c3fe8b is lovely!